library/seq.lisp

```diff
@@ -0,0 +1,517 @@
+                 (make-node-from-branches taller-branches))))))
+
+  (define (copy seq)
+    "A shallow copy of seq"
+    (match seq
+      ((LeafArray vec)
+       (LeafArray (vector:copy vec)))
+      ((RelaxedNode ht fss cst seq)
+       (RelaxedNode ht fss (vector:copy cst) (vector:copy seq)))))
+  
+  (declare rebalance-branches (types:runtimerepr :a => vector:Vector (Seq :a) -> Seq :a))
+  (define (rebalance-branches branches)
+    "Ensures each member of BRANCHES has between MIN-BRANCHING and
+MAX-BRANCHING subnodes of their own. Returns a new Seq whose branches
+are the rebalanced BRANCHES. 
+
+This function assumes that each member of BRANCHES has the same
+height. It also assumes that each member of BRANCHES is itself already
+balanced.
+
+It attempts to rebalance with a minimum of array copying."
+    (let ((stop
+            (- (vector:length branches) 1))
+          (cached-branch
+            (cell:new None))
+          (branch-rebalancer
+            (compose
+             rebuild-size-table
+             (fn (i)
+               (let branch =
+                 (match (cell:read cached-branch)
+                   ((None) (vector:index-unsafe i branches))
+                   ((Some cached) cached)))
+               (let subbranch-count = (branch-count branch))
+               (cond
+                 ((or (== i stop) (<= min-branching subbranch-count))
+                  (cell:swap! cached-branch None)
+                  branch)
+                 (True
+                  (let this-branch =
+                    (if (optional:some? (cell:read cached-branch))
+                        branch          ; already a copy
+                        (copy branch)))
+                  ;; need to mutate the next branch so we copy it
+                  (let next-branch = 
+                    (copy (vector:index-unsafe (+ i 1) branches)))
+                  ;; Do the subbranch shifting. NOTE: the choice of
+                  ;; branch capacity comes with trade-offs: using
+                  ;; MAX-BRANCHING tends to keep trees short over
+                  ;; time, improving lookup speed. Using MIN-BRANCHING
+                  ;; tends to minimize the number of array copies we
+                  ;; will make over time. I have chosen MIN-BRANCHING.
+                  (%shift-n-branches-onto
+                   this-branch next-branch (- min-branching subbranch-count))
+                  ;; cache the branch for the next round  
+                  (cell:swap! cached-branch (Some next-branch))
+                  this-branch)))))) 
+
+      ;;make a node from rebalanced branches
+      (make-node-from-branches
+       (iter:collect!
+        (iter:filter! (compose (< 0) size)
+                      (map branch-rebalancer
+                           (iter:up-through stop))))))))
+
+(cl:defmacro make (cl:&rest elems)
+  "Create a new Seq containing ELEMS."
+  `(fold push (new) (make-list ,@elems)))
```

Comment:
I would avoid consing the list. Maybe just expand into a chain of PUSH, to match how MAKE-LIST works?